// @ombiel/eslint-config-cm/index.js
module.exports = {
  parser: "babel-eslint",
  env: {
    browser: true,
    node: true,
    jest: true,
    es6: true,
  },

  plugins: [
    "react",
  ],

  parserOptions: {
    ecmaVersion: 6,
    sourceType: "module",
    ecmaFeatures: {
      jsx: true
    }
  },

  extends: "airbnb",

  // // This has started causing issues and is not essential
  // // for general module resolution in the new AEK boilerplate

  // settings: {
  //   "import/resolver": "webpack",
  // },

  rules: {
    "arrow-body-style": 0,
    "arrow-parens": 0,
    "arrow-spacing": 0,
    "brace-style": [1, "stroustrup",{ allowSingleLine: true }],
    "class-methods-use-this": 0,
    "comma-dangle": 0,
    "comma-spacing": 0,
    "consistent-return": 1,
    curly: 2,
    eqeqeq: [2, "allow-null"],
    "func-names": 0,
    "global-require": 0,
    "import/imports-first": 0,
    "import/newline-after-import": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "import/no-named-as-default": 0,
    "import/no-unresolved": 2,
    "import/no-webpack-loader-syntax": 0,
    "import/prefer-default-export": 0,
    indent: [2,2,{
      SwitchCase: 1, 
      MemberExpression: 0, 
      ignoredNodes: ["TemplateLiteral > *"]
    }],
    "jsx-a11y/aria-props": 1,
    "jsx-a11y/heading-has-content": 0,
    // "jsx-a11y/href-no-hash": 2,
    "jsx-a11y/label-has-for": 1,
    "jsx-a11y/mouse-events-have-key-events": 1,
    "jsx-a11y/role-has-required-aria-props": 1,
    "jsx-a11y/role-supports-aria-props": 1,
    "key-spacing": 1,
    "max-len": 0,
    "new-cap": 2,
    "newline-per-chained-call": 0,
    "no-confusing-arrow": 0,
    "no-console": 1,
    "no-loop-func": 2,
    "no-multiple-empty-lines": 0,
    "no-param-reassign": 0,
    "no-trailing-spaces": 0,
    "no-underscore-dangle": 0,
    "no-unused-expressions": 2,
    "no-unused-vars": [2,{ignoreRestSiblings: true}],
    "no-use-before-define": 2,
    "no-useless-rename": [1,{ignoreDestructuring: true}], //destructuring seems to be broken with spread assignments
    "no-var": 0,
    "no-with": 2,
    "object-curly-newline": ["error", {
      ObjectExpression: {consistent: true, multiline: true},
      ObjectPattern: {consistent: true, multiline: true},
      ImportDeclaration: {consistent: true, multiline: true},
      ExportDeclaration: {consistent: true, multiline: true},
    }],
    "object-curly-spacing": 0,
    "object-shorthand": 0,
    "padded-blocks": 0,
    "prefer-destructuring": 0,
    "react/destructuring-assignment": 0,
    "prefer-template": 1,
    "prefer-arrow-callback": 0,
    quotes: 0,
    "react/forbid-prop-types": 0,
    "react/jsx-filename-extension": 0,
    "react/jsx-first-prop-new-line": [2,"multiline"],
    "react/jsx-no-duplicate-props": 2,
    "react/jsx-no-target-blank": 0,
    "react/jsx-no-undef": 2,
    "react/jsx-one-expression-per-line": 0,
    "react/jsx-uses-react": 2,
    "react/require-default-props": 0,
    "react/require-extension": 0,
    "react/self-closing-comp": 0,
    "require-yield": 0,
    semi: 2,
    "space-before-function-paren": 0,
    "spaced-comment": 0,
    "space-infix-ops": 2,
    strict: 0,
    "vars-on-top": 0,
  },
};
