// @ombiel/aek-lib/lib/utils.js

export {
  buildParams,
  fuse,
  plate,
  param,
  parseQueryString,
  getWindowDimensions,
  outerHeight,
  outerWidth,
  isReactComponent,
  normaliseError,
  asArray,
  validateEmail,
  validateTel,
  switchCase,
  titleCase,
} from "@ombiel/cm-lib/lib/utils";


// aek environment helpers
export function getEnvironment() {
  return window.document.documentElement.getAttribute("data-env");
}

export function getEnv() {
  return getEnvironment();
}

export function isDev() {
  return getEnv() === "dev";
}

export function getPlatform() {
  return window.document.documentElement.getAttribute("data-platform") || "";
}

export function isWeb() {
  return getPlatform().substring(0,3).toLowerCase() === "web";
}

export function getPackageData() {
  var json = window.document.documentElement.getAttribute("data-aek-package");
  if (!json) {
    return null;
  }
  try {
    var data = JSON.parse(json);
    return data;
  }
  catch (e) {
    return null;
  }
}

export function getPackageName() {
  var packageData = getPackageData();
  if (packageData) {
    return packageData.name;
  }
  return null;
}

export function isNative() {
  return !isWeb();
}

export function currentURL() {
  var currentLocation = document.body.getAttribute("data-location");
  // newURL = currentLocation
  // currentParams = aek.parseQueryString(currentLocation)
  // redirectBase = currentParams[redirectBaseKey]
  // params[redirectBaseKey] = null
  // newURL = aek.modifyQueryString(currentLocation,params)

  // if aek.isNative()
  //   if newActivity
  //     newURL = prefixSelfReferenceUrl(newURL,currentLocation)
  //   else if redirectBase
  //     newURL = prefixSelfReferenceUrl(newURL,currentLocation,true,redirectBase)
  // return newURL

  if (!currentLocation) {
    currentLocation = window.location.pathname;
  }

  return currentLocation;
}

export function currentAbsoluteURL() {
  var currentLocation = currentURL();
  return `${window.location.protocol}//${window.location.host}${currentLocation}`;
}

export function openURL(url) {
  var customURL;
  try {
    customURL = window.top.CampusM.page.customURL;
  }
  catch (e) {} //eslint-disable-line no-empty

  if (customURL) {
    customURL = window.top.CampusM.page.customURL(url);
  }

  if (!customURL) {
    window.open(url);
  }
}

export function getBodyDataAttribute(key) {
  var obj = window.document.body.getAttribute(`data-o-b-j-${key}`);
  if (obj) {
    return JSON.parse(obj);
  }
  return window.document.body.getAttribute(`data-${key}`);

}

export function publicPath(path) {
  var normPath = path.replace(/^(\/public)?\//gi,"");
  var pPath = window.document.documentElement.getAttribute("data-public-assets");

  if (pPath) {
    return `${pPath}/${normPath}`;
  }

  if (isDev()) {
    return `/public/${normPath}`;
  }
  var pkg = getPackageData();
  return `https://portal.ombiel.co.uk/assets/aek/${pkg.name}/${pkg.version}/${normPath}`;
}
