// @ombiel/aek-devserver/index.js
const Path = require("path");
const _ = require("lodash");
const chokidar = require("chokidar");
const webpack = require("webpack");
const WebpackDevServer = require("webpack-dev-server");
const getWebpackConfig = require("webpack-cli/bin/convert-argv");
const log = require("@ombiel/aek-cli-support/logger");
const chalk = require("chalk");

const auth = require("@ombiel/aek-cli-support/auth");

const {getUserDetails} = auth;
const initBeforeMiddleware = require("./middleware/before-middleware");
const initAfterMiddleware = require("./middleware/after-middleware");

function aekDevServer(options = {}) {

  const {port} = options;
  let server;

  const srcBase = options.srcBase || "./src/client";

  let webpackConfig = getWebpackConfig({_: []});
  let entry;

  function restart() {

    if (server) { server.close(); }

    return getUserDetails(options)
    .then((userDetails)=>{
      options.userDetails = userDetails;

      return initBeforeMiddleware(options)
      .then((beforeMiddleware)=>{

        const afterMiddleware = initAfterMiddleware(options);

        server = new WebpackDevServer(webpack(webpackConfig),{
          before: (app)=>app.use(beforeMiddleware),
          after: (app)=>app.use(afterMiddleware),
          publicPath: "/public",
          stats: webpackConfig.stats,
          disableHostCheck: true,
        });

        return server.listen(port,null,()=>{
          log.info(`aek devserver running at ${chalk.cyan(`http://localhost:${port}`)} targetting ${chalk.blue(options.hostname)}`);
        });
      });
    });
  }

  function retry() {
    if (_.isFunction(webpackConfig.then)) {
      webpackConfig.then((resolved)=>{
        webpackConfig = resolved;
        retry();
      });
      return;
    }

    if (!_.isEqual(webpackConfig.entry,entry)) {
      restart();
    }
    entry = webpackConfig.entry;

  }

  retry();

  function onChange() {
    webpackConfig = getWebpackConfig({_: []});
    retry();
  }

  chokidar.watch(Path.resolve(srcBase))
  .on("add",onChange)
  .on("unlink",onChange);


}

module.exports = aekDevServer;
