// @ombiel/aek-lib/react/router/mem-location.es6.js
var LocationActions = require('react-router/modules/actions/LocationActions');
// var getWindowPath = require('../utils/getWindowPath');

var _actionType,
    history = ["/"],
    _onChange;


function getPath() {
  return history[history.length-1];
}

function ensureSlash() {
  var path = getPath();

  if (path.charAt(0) === '/')
    return true;

  MemLocation.replace('/' + path);

  return false;
}

function updateLocation() {
  if (ensureSlash()) {
    var path = getPath();

    _onChange({
      // If we don't have an _actionType then all we know is the hash
      // changed. It was probably caused by the user clicking the Back
      // button, but may have also been the Forward button or manual
      // manipulation. So just guess 'pop'.
      type: _actionType || LocationActions.POP,
      path: getPath()
    });

    _actionType = null;
  }
}

/**
 * A Location that uses page memory.
 */
var MemLocation = {

  setup: function (onChange) {
    _onChange = onChange;

    // Do this BEFORE listening for hashchange.
    // ensureSlash();

    // if (window.addEventListener) {
    //   window.addEventListener('hashchange', onHashChange, false);
    // } else {
    //   window.attachEvent('onhashchange', onHashChange);
    // }
  },

  teardown: function () {
    // if (window.removeEventListener) {
    //   window.removeEventListener('hashchange', onHashChange, false);
    // } else {
    //   window.detachEvent('onhashchange', onHashChange);
    // }
  },

  push: function (path) {
    _actionType = LocationActions.PUSH;
    history.push(path)
    updateLocation()
  },

  replace: function (path) {
    _actionType = LocationActions.REPLACE;
    history = history.slice(0,history.length-1).concat([path])
    updateLocation();
  },

  pop: function () {
    _actionType = LocationActions.POP;
    history.pop();
    updateLocation();
  },

  getCurrentPath: getPath,

  toString: function () {
    return '<MemLocation>';
  }

};

module.exports = MemLocation;
