// @ombiel/aek-lib/simple-store/index.es6.js
var EventEmitter = require("-aek/event-emitter");
var _ = require("-aek/utils");

var Immutable = require("seamless-immutable");

var rePropName = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\n\\]|\\.)*?)\2)\]/g;

/**
 * Converts `value` to property path array if it's not one.
 *
 * @private
 * @param {*} value The value to process.
 * @returns {Array} Returns the property path array.
 */
function toPath(value) {
  if (_.isArray(value)) {
    return value;
  }
  var result = [];
  value.replace(rePropName, function(match, number, quote, string) {
    result.push(quote ? string.replace(/\\(\\)?/g, '$1') : (number || match));
  });
  return result;
}

// remove circular refs and functions
function clean(obj,refs = []) {
  if(!obj) { return obj; }
  if(Immutable.isImmutable(obj)) { return obj; }
  if(typeof obj === "function") {
    return "[Function]";
  }
  if(typeof obj === "object") {
    if(_.includes(refs,obj)) {
      return "[Circular]";
    }
    refs = refs.concat([obj]);

    if(_.isArray(obj)) {
      return obj.map((item,i)=>{
        return clean(item,refs);
      })
    }

    var cleaned = {};
    _.forEach(obj,(val,key)=>{
      cleaned[key] = clean(val,refs);
    });
    return cleaned;
  }

  return obj;
}

class StoreContext {

  constructor(store,base) {
    this.store = store;
    this.base = base;
    this.promise = Promise.resolve();
  }

  dispatch(desc) {
    if(desc.error) { desc = _.extend({name:"error"},desc); }
    return this.store.dispatch(_.extend({},this.base,desc));
  }

}

class SimpleStore extends EventEmitter {

  constructor(opts) {

    opts = _.extend({plugins:[],initialState:{}},opts);

    super();

    this.state = Immutable.from(opts.initialState);

    this.constructor.plugins.forEach((plugin)=>{
      this.plugin(plugin);
    });

    opts.plugins.forEach((plugin)=>{
      this.plugin(plugin);
    });

    this.prevState = this.state;

    this.emit("init",this.state);

  }

  asMutable(obj) {
    if(obj && obj.asMutable) {
      return Immutable.asMutable(obj,{deep:true});
    }
    return obj;
  }

  checkChange() {
    if (_.isEqual(this.prevState,this.state)) {
      this.emit("nochange",this.state);
    }
    else {
      this.emit("change",this.state,this.prevState);
    }
    this.prevState = this.state;
  }

  dispatch(desc) {
    this.emit("dispatch",desc);

    let {group,name,path,payload,replace,extend,...opts} = desc;

    if(!(group || name)) { throw Error("Dispatch must include a group or name"); }
    if(!(_.isString(path) || _.isArray(path))) { throw Error("Dispatch must include a valid path"); }
    if(!_.has(desc,"replace") && !_.has(desc,"extend")) { throw Error("Dispatch must include a modifier (extend or replace)"); }

    if (extend && (replace == null)) {
      if(_.isFunction(extend)) {
        replace = (state,payload)=>_.extend(state,extend(state,payload));
      }
      else {
        replace = ()=>_.extend({},this.get(path),extend);
      }
    }

    if(!_.isFunction(replace)) {
      let v = replace;
      replace = ()=>v;
    }

    var newState,val;

    if (replace.length) {
      val = replace(this.asMutable(this.get(path)),payload);
    }

    else {
      val = replace();
    }

    // clean payload
    var cleaned = clean(val);

    if(!path) {
      newState = Immutable.replace(this.state,cleaned,{deep:true});
    }
    else {
      newState = Immutable.setIn(this.state,toPath(path),cleaned,{deep:true});
    }

    this.state = newState;
    this.checkChange();

  }

  context(base) {
    return new StoreContext(this,base);
  }

  get(path,defaultValue) {
    if(!path) {
      return this.state || Immutable.from(defaultValue);
    }
    return _.get(this.state,path,Immutable.from(defaultValue));
  }

  set(path,val) {
    val = clean(val);
    this.emit("set",path,val);
    if(!path) {
      this.state = Immutable.replace(this.state,val,{deep:true});
    }
    else {
      this.state = Immutable.setIn(this.state,toPath(path),val,{deep:true});
    }
    this.checkChange();
  }

  plugin(plugin) {
    plugin(this);
  }


}

SimpleStore.plugins = [];

module.exports = SimpleStore;
