// @ombiel/aek-lib/notifier.jsx
var NotificationSystem = require("react-notification-system");
// var STYLES = require("react-notification-system/src/styles");
var React = require("react");
var ReactDOM = require("react-dom");
var _ = require("./utils");

var defaultNotification = {
  title: null,
  message: null,
  level: "info",
  position: 'bc',
  autoDismiss: 5,
  dismissible: true,
  action: null
};


var notifications;

function init() {
  var el = document.createElement("div");
  el.style.position = "absolute";
  document.body.appendChild(el);
  notifications  = ReactDOM.render(<NotificationSystem/>,el);
}

if(document.body) { init(); }
else {
  document.addEventListener('DOMContentLoaded',init);
}

var Notifier = {
  add(notification) {
    return notifications.addNotification(_.extend({},defaultNotification,notification));
  },
  remove(notification) {
    return notifications.removeNotification(notification);
  },
  clear() {
    return notifications.clearNotifications();
  }
};

module.exports = Notifier;
